src/main/java/net/minestom/server/instance/block/predicate/CollectionPredicate.java

```diff
@@ -0,0 +1,97 @@
+package net.minestom.server.instance.block.predicate;
+
+import net.minestom.server.codec.Codec;
+import net.minestom.server.codec.StructCodec;
+import net.minestom.server.utils.Range;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+
+public record CollectionPredicate<T, P extends Predicate<T>>(Contains<T, P> contains, Count<T, P> counts,
+                                     Range.Int size) implements Predicate<Iterable<T>> {
+
+    public static <T, P extends Predicate<T>> Codec<CollectionPredicate<T, P>> createCodec(Codec<P> codec) {
+        return StructCodec.struct(
+                "contains", Contains.createCodec(codec).optional(), CollectionPredicate::contains,
+                "count", Count.createCodec(codec).optional(), CollectionPredicate::counts,
+                "size", DataComponentPredicates.INT_RANGE_CODEC.optional(), CollectionPredicate::size,
+                CollectionPredicate::new
+        );
+    }
+
+    @Override
+    public boolean test(Iterable<T> collection) {
```

Comment:
From what I can tell, the data components that use this structure (`attribute_modifiers`, `bundle_contents`, `container`, `fireworks`, `writable_book_content`, `written_book_content`) all store the items as lists, so this should be able to be converted to a `List<T>` or `Collection<T>` which means you should be able to work over one of those types and remove the ugly `sizeOf` method and some other associated stuff. LMK if I missed something that makes this impossible, though.

Reply:
Okay I can see it used in `AttributeModifiers#test` but I think this isn't at all a hard dependency.